sfusurge/hacker-portal
# Basic test setup

## Describe your changes

Setup vitest and playwright for unit and e2e tests. Unit tests use a real database that gets created before test is run and deleted after they finished. Before each test case, all tables are cleaned so that they have a clean state to work with.

Tests are organized in the `test/<unit | e2e>` using vitest's workspace feature

Added two new npm commands: `npm run test:unit` and `npm run test:e2e`

## Testing steps

`npm run test:unit` output:

```
> portal@0.1.0 test:unit
> vitest run --project unit

The CJS build of Vite's Node API is deprecated. See https://vite.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.

 RUN  v2.1.5 /Users/scott/workspace/hacker-portal

Setting up test database
[dotenv@16.4.5][DEBUG] No encoding is specified. UTF-8 is used by default
Loaded config {
    "host": "localhost",
    "port": 3306,
    "user": "scott",
    "password": "<redacted>",
    "database": "surge_test"
}
> drizzle-kit push
 ✓ |unit| test/unit/hackathon.test.ts (6)
 ✓ |unit| test/unit/user.test.ts (4)

 Test Files  2 passed (2)
      Tests  10 passed (10)
   Start at  13:59:55
   Duration  756ms (transform 27ms, setup 63ms, collect 85ms, tests 146ms, environment 0ms, prepare 41ms)

Dropped database surge_test
```

`npm run test:e2e` (have `npm run dev` in another window) output:

```
> portal@0.1.0 test:e2e
> vitest run --project e2e

The CJS build of Vite's Node API is deprecated. See https://vite.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.

 RUN  v2.1.5 /Users/scott/workspace/hacker-portal

Setting up test database
[dotenv@16.4.5][DEBUG] No encoding is specified. UTF-8 is used by default
Loaded config {
    "host": "localhost",
    "port": 3306,
    "user": "scott",
    "password": "<redacted>",
    "database": "surge_test"
}
> drizzle-kit push
 ✓ |e2e| test/e2e/basic.test.ts (1) 2538ms
   ✓ basic e2e tests (1) 2538ms
     ✓ basic 2038ms

 Test Files  1 passed (1)
      Tests  1 passed (1)
   Start at  14:01:07
   Duration  3.20s (transform 14ms, setup 35ms, collect 250ms, tests 2.54s, environment 0ms, prepare 20ms)

Dropped database surge_test
```

## Issue ticket number and link

## Checklist before requesting a review
